OpenRA.Mods.Common/Activities/Air/Land.cs

```diff
@@ -48,6 +48,17 @@ public override Activity Tick(Actor self)
 
 			if (IsCanceling || target.Type == TargetType.Invalid)
 			{
+				// Make aircraft climb back to CruiseAltitude before moving on to NextActivity or going idle.
+				// TODO: Try to fix that even if NextActivity is not null, VTOLs will not turn towards the target while climbing,
```

Comment:
I think this will be best handled by having those activities opt out of the altitude fixup here and handle the taking off themselves (which they are almost certainly already doing).

Reply:
Generally, they don't. They would if we queue `TakeOff` in `Fly.OnFirstRun` whenever the altitude is lower than `Aircraft.CruiseAltitude`, as I already suggested.

The only aircraft related activity that takes care of this itself, as far as I can see, is `HeliAttack` and it really shouldn't. All these activities should just queue `Fly` instead of calling `Fly.FlyTick` or any other internal `Fly` method directly. `Fly` would then queue `TakeOff` as necessary.

Reply:
Makes sense. Replacing `FlyTick` with `Fly` children is going to be important for the acceleration and cliff height prediction logic too.

The activities that queue `Fly` directly or use `move.Move*` will then be able to tag themselves with `IActivityAircraftAltitudeAware` (or whatever we want to call it) to opt out of the internal `Land` handling.

Reply:
In hindsight, my request to allow facing changes has caused more trouble than it's worth. In practice it also looks a bit odd to have units turn before landing on a helipad but then turn while taking off when reloading is complete.

How about we abandon this idea and keep it simple?

Reply:
I think units should be able to turn while landing as well.

 Since we will probably want to replace `FlyTick` with `Fly` children anyway and `Fly` is already queueing `TakeOff` (but in the wrong place), this _is_ keeping it simple.

Also, since any activity that cares about positioning at all will be queueing either `Fly` directly or using the `IMove` interface, we will be better off making the the "`TakeOff` on cancelling `Land`" thing an opt-in instead of an opt-out.

 (In fact, I've got the sneaking suspicion we will be left with exactly zero activities that would actually benefit from this, so we can just leave it for the idling case, making it even more simple.)

Reply:
> I think units should be able to turn while landing as well.

Turning when landing is more complicated because we need to make sure that the facing is correct before it touches down, which may mean turning part way before starting to descend.

> Also, since any activity that cares about positioning at all will be queueing either Fly directly or using the IMove interface, we will be better off making the "TakeOff on cancelling Land" thing an opt-in instead of an opt-out.

Pragmatically, I expect you're correct. I'm just a bit annoyed that we still have to hardcode exceptions for aircraft in places like `Transforms`.

Reply:
My main motivation for suggesting we drop the facing changes during takeoff is that the fly code is too broken to tackle in one PR, and each time through review we seem to find another direction or set of issues to scope creep it on. This is an area where we can at least keep the previous behaviour without introducing anything new to complain about, then the behaviour on both takeoff and land can be addressed together in a later PR.

Reply:
I agree that endlessly debating this isn't getting us anywhere. We should probably leave this as is and get this and #16509 merged quickly, so we can tackle these loose ends in one PR. Throwing away the facing changes during take-off would be a mistake though. Better to keep what we have here, so we have something to work from and polish later.

Reply:
I've been assuming that turning off the facing adjustment can be done by removing the `var desiredFacing = delta.HorizontalLengthSquared != 0 ? delta.Yaw.Facing : aircraft.Facing;` line and calling `VerticalTakeOffOrLandTick` with `aircraft.Facing` instead. This can be trivially restored in the followup.

Reply:
You're right. If it is just the one line, then that isn't a big deal.